ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -590,140 +590,286 @@ private boolean copyDataByIndexSegment(LogSegment logSegmentToCopy, IndexSegment
+          if (secondVal != null && secondVal.isFlagSet(IndexValue.Flags.Ttl_Update_Index) && isTtlUpdateEntryValid(
+              indexEntry.getKey(), indexSegment.getStartOffset())) {
+            validEntries.add(new IndexEntry(indexEntry.getKey(), secondVal));
           }
+        } else {
+          // if this delete cannot be counted and there is a corresponding unexpired PUT/TTL update entry in the same
+          // index segment, we will need to add it.
+          addAllEntriesForKeyInSegment(validEntries, indexSegment, indexEntry);
+        }
+      } else if (value.isFlagSet(IndexValue.Flags.Ttl_Update_Index)) {
+        // if IndexSegment::getIndexEntriesSince() returns a TTL update entry, it is because it is the ONLY entry i.e.
+        // no PUT or DELETE in the same index segment.
+        if (isTtlUpdateEntryValid(indexEntry.getKey(), indexSegment.getStartOffset())) {
+          validEntries.add(indexEntry);
         }
-      } else if (!srcIndex.isExpired(value)) {
-        // unexpired PUT entry.
-        if (deletesInEffect) {
-          FileSpan deleteSearchSpan =
-              new FileSpan(indexSegmentStartOffset, startOffsetOfLastIndexSegmentForDeleteCheck);
-          IndexValue searchedValue = srcIndex.findKey(indexEntry.getKey(), deleteSearchSpan,
-              EnumSet.of(PersistentIndex.IndexEntryType.PUT, PersistentIndex.IndexEntryType.DELETE));
-          if (!searchedValue.isFlagSet(IndexValue.Flags.Delete_Index)) {
-            // PUT entry that has not expired and is not considered deleted.
-            validEntries.add(indexEntry);
+      } else {
+        IndexValue valueFromIdx = srcIndex.findKey(indexEntry.getKey());
+        // Doesn't matter whether we get the PUT or DELETE entry for the expiry test
+        if (!srcIndex.isExpired(valueFromIdx)) {
+          // unexpired PUT entry.
+          if (deletesInEffect) {
+            if (!hasDeleteEntryInSpan(indexEntry.getKey(), indexSegment.getStartOffset(),
+                startOffsetOfLastIndexSegmentForDeleteCheck)) {
+              // PUT entry that has not expired and is not considered deleted.
+              // Add all values in this index segment (to account for the presence of TTL updates)
+              addAllEntriesForKeyInSegment(validEntries, indexSegment, indexEntry);
+            } else {
+              logger.trace("{} in index segment with start offset {} in {} is not valid because it is a deleted PUT",
+                  indexEntry, indexSegment.getStartOffset(), storeId);
+            }
           } else {
-            logger.trace("Skipping {} in index segment with start offset {} in {} because it is a deleted PUT",
-                indexEntry, indexSegmentStartOffset, storeId);
+            // valid PUT entry
+            // Add all values in this index segment (to account for the presence of TTL updates)
+            addAllEntriesForKeyInSegment(validEntries, indexSegment, indexEntry);
```

Comment:
If `oneEntryPerKey`, `PUT` should be the latest and only one in the `indexSegment`. How `TTL` presents in this `indexSegment` ?

Reply:
If you look at the code in `IndexSegment#eliminateDuplicates()`, these are the possible input and output combinations
1. PUT entry only -> PUT entry
2. TTL update entry only -> TTL update entry
3. DELETE entry only -> DELETE entry
4. PUT + DELETE -> DELETE
5. TTL update + DELETE -> DELETE
6. PUT + TTL update -> PUT (the one relevant to this comment)
7. All three -> DELETE

Reply:
Added a comment saying that this (^) is assumed